Fix creature wince animation for low morale and lich cloud cases. Battlefield.

Districh-ru opened the pull request:
close #6367

Low morale effect:

https://user-images.githubusercontent.com/113276641/209448488-5a1ddd39-03c4-48b3-9989-235df96cb867.mp4

Lich attack:

https://user-images.githubusercontent.com/113276641/209448487-398c785c-08aa-495f-b6ca-ae49db07c3a0.mp4

oleg-derevenetz commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -5450,52 +5468,57 @@ void Battle::Interface::RedrawTargetsWithFrameAnimation( const TargetsInfo & tar
 }
 
-void Battle::Interface::RedrawTroopWithFrameAnimation( Unit & b, int icn, int m82, CreatueSpellAnimation animation )
+void Battle::Interface::RedrawTroopWithFrameAnimation( Unit & unit, int icn, int m82, CreatueSpellAnimation animation )
```
Hi @Districh-ru I don't know when exactly this type was introduced, but there is definitely a typo: `CreatueSpellAnimation` -> `CreatureSpellAnimation`. Also please rename the first parameter in this method's declaration in the `battle_interface.h`: `b` -> `unit`.

Districh-ru replied:
Hi, @oleg-derevenetz. Thank you for your attention. Fixed this.

oleg-derevenetz reviewed (approved):
Hi @Districh-ru I just left one optional suggestion, LGTM in general.

Districh-ru commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -3260,50 +3313,63 @@ void Battle::Interface::RedrawActionWincesKills( const TargetsInfo & targets, Un
+        AudioManager::PlaySound( attacker->M82Expl() );
+    }
+
+    uint32_t frame = 0;
```
Hi, @oleg-derevenetz. Yes, for now we have only this ICN to render at this place.

ihhub commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -2989,6 +2989,64 @@ void Battle::Interface::MouseLeftClickBoardAction( int themes, const Cell & cell
     }
 }
 
+fheroes2::Point CalculateSpellPosition( const Battle::Unit & target, int spellICN, const fheroes2::Sprite & spellSprite )
```
Since we are moving this function across the file please move it to anonymous namespace at the top of the file, line 306. Right now this function has global scope which is not recommended.

ihhub commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -3260,50 +3313,63 @@ void Battle::Interface::RedrawActionWincesKills( const TargetsInfo & targets, Un
 
+    // If this was a Lich attack, we should render an explosion cloud over the target unit immediately after the projectile hits the target,
+    // along with the unit kill/wince animation.
+    const bool drawLichCloud = attacker != nullptr && defender != nullptr && attacker->isArchers() && !attacker->isHandFighting()
```
Can you please add parenthesis over `attacker != nullptr` and `defender != nullptr` for easier readability?
```cpp
( attacker != nullptr ) && ( defender != nullptr )
```

ihhub commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -3260,50 +3313,63 @@ void Battle::Interface::RedrawActionWincesKills( const TargetsInfo & targets, Un
         if ( Game::validateAnimationDelay( Game::BATTLE_FRAME_DELAY ) ) {
-            if ( finishedAnimation ) {
-                // All frames are rendered.
+            if ( finishedAnimation && ( !drawLichCloud || ( drawLichCloud && lichCloudFrame == fheroes2::AGG::GetICNCount( ICN::LICHCLOD ) ) ) ) {
```
The second condition in this line can be simplified:
```cpp
if ( finishedAnimation && ( !drawLichCloud || ( lichCloudFrame == fheroes2::AGG::GetICNCount( ICN::LICHCLOD ) ) ) ) {
```
as `drawLichCloud` will be true for the last condition.

Districh-ru replied:
Hi @ihhub, I've also cached `fheroes2::AGG::GetICNCount( ICN::LICHCLOD )` before the loop and moved `PlaySound` close to the loop.

ihhub replied:
Good job!

ihhub reviewed (changes requested):
Hi @Districh-ru , I put few comments in this pull request. Could you please take a look when you have time?

Branikolog commented:
Hi, @Districh-ru 
During cloud attack creatures seem to have a small pause in the middle of wince animation, same to low morale animation. Not sure it looks better, than current PR implementation though.
Not sure, it's related to current PR, but animation after gold eagle disappeared for some creature seem to be too prompt. For instance you can barely notice few animation frames before Trolls or Crusaders return to their normal state. But at the same time Pikemen look OK here...
Anyway, now pause is made at correct time. If this was the issue this PR fixes, than we can merge.

Districh-ru commented:
> Hi, @Districh-ru During cloud attack creatures seem to have a small pause in the middle of wince animation, same to low morale animation. Not sure it looks better, than current PR implementation though. 

The wince animation speed here is equal to animation of Lich cloud or Golden Eagle. If it is needed we can make creatures to wait for 1 frame, but this will complicate the render logic a bit.

> Not sure, it's related to current PR, but animation after gold eagle disappeared for some creature seem to be too prompt. For instance you can barely notice few animation frames before Trolls or Crusaders return to their normal state. But at the same time Pikemen look OK here...

We have fixed the issue with lost frames in #6398. In that fix I also added the check for delay after Golden eagle animation and before next creatures turn. So let's check this issue again after #6398 PR merge. :)

Branikolog commented:
Hi, @Districh-ru !

>We have fixed the issue with lost frames in https://github.com/ihhub/fheroes2/pull/6398. In that fix I also added the check for delay after Golden eagle animation and before next creatures turn. So let's check this issue again after https://github.com/ihhub/fheroes2/pull/6398 PR merge. :)

Good to hear!

>The wince animation speed here is equal to animation of Lich cloud or Golden Eagle. If it is needed we can make creatures to wait for 1 frame, but this will complicate the render logic a bit.

There's a pause during low morale animation in current PR implementation which was during cloud attack in the OG too.
The same pause occurs during wave/shout spell too. 

https://user-images.githubusercontent.com/55348946/211026128-60af93b5-2c0f-4d84-9693-ad0ed7c58b74.mp4

Districh-ru commented:
Hi, @Branikolog.
You mean the pause after you skip the turn and before the low morale eagle appears?
I've checked these pauses and cloud not reproduce the issue. Maybe because I have a powerful CPU.
Could you please check if this pauses occurs in current master build. And also the release v1.0. 
It seems that some calculations are made and the game waits for them to finish ... And if it is so, we have to optimize that part of code.
If this reproduces in master build then we have to open a separate issue as this PR, as well as #6452 changed only the animation logic and did not touch the game logic outside the current animation.

Branikolog commented:
@Districh-ru 
I mean pause during wince animation. On my last video crusader winces back, holding that pose for a while and then returns to it normal state. The same pause during wince animation was during cloud appears in the OG. In current PR build troops under the cloud show full wincing animation at once.
Sorry for not being clear. :)

Districh-ru commented:
> @Districh-ru I mean pause during wince animation. On my last video crusader winces back, holding that pose for a while and then returns to it normal state. The same pause during wince animation was during cloud appears in the OG. In current PR build troops under the cloud show full wincing animation at once. Sorry for not being clear. :)

@Branikolog, thanks for clearing! :) 
I made this animation without the pause because I took a video of OG only for one shot and there the zombie does not pause in the maximum wince state:

https://user-images.githubusercontent.com/113276641/211036486-1cc3a4cd-4ef4-49a3-b5fb-3bbd0ffc562d.mp4

Maybe OG has bugs and some units have this pause and some - not.
I can try to add this pause. But as it is a blast I thought that the hit creature does not have to pause in the middle of wince. :)

Branikolog commented:
@Districh-ru 
On the original game it is probably made in such way to make all wince animations to last the same time.
Your current implementation looks well, but we have same wince pausing during wave/shout spells too. So I think animation could be consistent with original one. Also, lich launches a small sphere and it looks like damage to the target crowd is made by the explosion. So, holding wincing animation for a little more time seems more natural.

Districh-ru commented:
> @Districh-ru On the original game it is probably made in such way to make all wince animations to last the same time. Your current implementation looks well, but we have same wince pausing during wave/shout spells too, if I'm not mistaken. So I think animation could be consistent with original one here. Also, lich launches a small sphere and it looks like damage to the target crowd is made by the explosion. So, holding wincing animation for a little more time seems more natural.

@Branikolog, OK.
The other animations with the pause do "wince up" animation an then waits for spell/morale animation to finish and then do the "wince down" animation.
The Lich cloud has 10 animation frames. the maximum wince animation has Halfling, Dwarf and Pikeman - 6 frames. So I plan to add a pause for 4 frames so every creature will end its wince animation before the cloud finishes its animation. What do you think?

Branikolog commented:
@Districh-ru 
I think both will look good. I'm only afraid of making pause too long, as creatures would look weirdly standing for too long period of time in that poses. 
I can tell you, I wouldn't have noticed this issue, if cloud animation had less frames or creature wincing animation started a few frames later. Not sure, it would look better, but I think a solution to avoid this wince pause could be waiting for a few frames after cloud animation started before starting wincing. Or even a combination of adding a lag and short wince pause. 
But I think, your suggestion will look maximum authentic.

Districh-ru commented:
@Branikolog, I've added the pause to wince animation. Please check it when you have time. We can change at what frame the "wince down" should start. Now, assuming that Lich bomb also has a gas, the "wince down" animation starts after the smoke cloud starts to disappear.


https://user-images.githubusercontent.com/113276641/211053222-483ed2c1-324d-4aab-ad0e-53238b9f9cc3.mp4

Branikolog commented:
@Districh-ru 
Looks great!

I'm thinking of making "wince down" 1 frame earlier, but I'm not sure, since I don't want to miss most part of those gorgeous animations under the cloud. :) Looks like peak of implosion happens on 5th frame, while current "wince down" starts on 8th. Some creatures have not so many frames, so, I'm not sure, we would notice that animation through the holes in the cloud.
Even without changing the frame it looks good and acceptable for me.

It's just my personal point.  Probably other enhancements could be done later, when we'd be polishing and improving animations beyond the original look. I personally would love to start first part of wincing animation at least 1-2 frames later, as most of creatures reach wincing pause long before implosion reaches its peak. They actually start wincing while implosion is so tiny and looks unharmful. If we come with this delay, I think current "wince down" timing will be even more suitable.

Districh-ru commented:
> It's just my personal point. Probably other enhancements could be done later, when we'd be polishing and improving animations beyond the original look. I personally would love to start first part of wincing animation at least 1-2 frames later, as most of creatures reach wincing pause long before implosion reaches its peak. They actually start wincing while implosion is so tiny and looks unharmful. If we come with this delay, I think current "wince down" timing will be even more suitable.

@Branikolog, I've added 1 frame pause before starting wince animation, as the second frame contains a small cloud and the creatures have a reason to start wincing.
And if we start the "wince down" animation one frame earlier, the most creatures, which have 1 frame of "wince down" animation, will finish it under the cloud.


https://user-images.githubusercontent.com/113276641/211133461-8352585e-1e6f-4b0d-b802-61e8d783de9a.mp4

Branikolog reviewed (approved):
It's perfect! Thanks, @Districh-ru !

ihhub commented on src/fheroes2/battle/battle_interface.cpp:
```diff
@@ -3282,43 +3334,68 @@ void Battle::Interface::RedrawActionWincesKills( const TargetsInfo & targets, Un
-            AudioManager::PlaySound( defender->M82Wnce() );
+            if ( drawLichCloud ) {
+                // The Lich cloud causes units to freeze for some time in the maximum wince state.
+                // So we will devide the wince animation. First part: the creature stands for couple frames before wincing.
```
Spelling correction: `devide` --> `divide`.

ihhub replied:
Small spelling correction: **for a couple of frames before wincing.**.

ihhub reviewed (commented):
Hi @Districh-ru , I put few recommendations about the comments. Could you please check them?

ihhub commented:
@Districh-ru , a bunch of thanks for this fix!
